src/jit/codegenxarch.cpp

```diff
@@ -1547,50 +1547,198 @@ void CodeGen::genCodeForJumpTrue(GenTreePtr tree)
 {
-    regNumber    dstReg      = setcc->gtRegNum;
-    CompareKind  compareKind = setcc->IsUnsigned() ? CK_UNSIGNED : CK_SIGNED;
-    emitJumpKind jumpKind    = genJumpKindForOper(setcc->gtCondition, compareKind);
+    regNumber dstReg = setcc->gtRegNum;
 
-    assert(genIsValidIntReg(dstReg) && isByteReg(dstReg));
-    // Make sure nobody is setting GTF_RELOP_NAN_UN on this node as it is ignored.
-    assert((setcc->gtFlags & GTF_RELOP_NAN_UN) == 0);
+    const GenConditionDesc& desc = GetConditionDesc(setcc->gtCondition);
 
-    inst_SET(jumpKind, dstReg);
-    inst_RV_RV(ins_Move_Extend(TYP_UBYTE, true), dstReg, dstReg, TYP_UBYTE, emitTypeSize(TYP_UBYTE));
+    if (desc.ins[1] == EJ_NONE)
+    {
+        inst_SET(desc.ins[0], dstReg);
+    }
+    else
+    {
+        inst_SET(desc.jmpToTrueLabel[0] ? desc.ins[0] : emitter::emitReverseJumpKind(desc.ins[0]), dstReg);
+        BasicBlock* skipLabel = genCreateTempLabel();
+        inst_JMP(desc.ins[0], skipLabel);
+        inst_SET(desc.ins[1], dstReg);
+        genDefineTempLabel(skipLabel);
+    }
+
+    getEmitter()->emitIns_R_R(INS_movzx, EA_1BYTE, dstReg, dstReg);
     genProduceReg(setcc);
 }
 
+//------------------------------------------------------------------------
+// genCodeForSelCC: Generate code for selecting a value based on a condition.
+//
+// Arguments:
+//    select - A GT_SELCC node
+//
+// Notes:
+//    CMOV is used if available, otherwise a conditional branch and a normal
+//    MOV is used instead.
+
+void CodeGen::genCodeForSelCC(GenTreeOpCC* selcc)
+{
+    assert(selcc->OperIs(GT_SELCC));
+
+    genConsumeOperands(selcc);
+
+    GenTree* op1 = selcc->gtGetOp1();
+    GenTree* op2 = selcc->gtGetOp2();
+
+    var_types srcType = op1->TypeGet();
+    var_types dstType = selcc->TypeGet();
+
+    assert(varTypeIsIntOrI(dstType));
+    assert(srcType == op2->TypeGet());
+    assert(op1->isUsedFromReg());
+    assert(op2->isUsedFromReg() || op2->isUsedFromMemory());
+
+    regNumber srcReg1 = op1->gtRegNum;
+    regNumber srcReg2 = op2->isUsedFromReg() ? op2->gtRegNum : REG_NA;
+    regNumber dstReg  = selcc->gtRegNum;
+
+    assert(genIsValidIntReg(srcReg1));
+    assert(op2->isUsedFromMemory() || genIsValidIntReg(srcReg2));
+    assert(genIsValidIntReg(dstReg));
+
+    if (dstReg != srcReg1)
+    {
+        inst_RV_RV(INS_mov, dstReg, srcReg1, dstType);
+        op1 = selcc;
+    }
+
+    if (compiler->opts.compUseCMOV)
+    {
+        const GenConditionDesc& desc = GetConditionDesc(selcc->gtCondition);
+        assert(desc.ins[1] == EJ_NONE);
+
+        constexpr instruction EJtoCMOV[]{INS_nop,    INS_nop,    INS_cmovo,  INS_cmovno, INS_cmovb,  INS_cmovae,
+                                         INS_cmove,  INS_cmovne, INS_cmovbe, INS_cmova,  INS_cmovs,  INS_cmovns,
+                                         INS_cmovpe, INS_cmovpo, INS_cmovl,  INS_cmovge, INS_cmovle, INS_cmovg};
+        noway_assert((unsigned)desc.ins[0] < _countof(EJtoCMOV));
```

Comment:
Would be nice to add comments with the respective EJ values here.

Reply:
I'm actually considering getting rid of the array, the more usual way this is done is something like `(kind - EJ_jo) + INS_cmovo`.